File: src/main/javabak/materials/Materials.java
package redcrafter07.processed.materials;

import net.minecraft.tags.BlockTags;
import net.minecraft.world.level.block.SoundType;
import net.minecraft.world.level.block.state.BlockBehaviour;
import net.minecraft.world.level.block.state.properties.NoteBlockInstrument;
import redcrafter07.processed.gui.RenderUtils;

import java.util.ArrayList;

public class Materials {
    public static ArrayList<Material> MATERIALS = new ArrayList<>();

    public static Material ALUMINIUM = register(new Material(
            "aluminium",
            RenderUtils.color(0xd0, 0xd5, 0xd9),
            "Al",
            BlockTags.NEEDS_STONE_TOOL,
            BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.BASEDRUM).strength(5, 10).requiresCorrectToolForDrops(),
            BlockBehaviour.Properties.of().sound(SoundType.STONE).strength(5, 10).requiresCorrectToolForDrops()
    ));
    public static Material STEEL = register(new Material(
            "steel",
            RenderUtils.color(0x49, 0x4b, 0x4d),
            "Fe",
            BlockTags.NEEDS_IRON_TOOL,
            BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.BASEDRUM).strength(5, 10).requiresCorrectToolForDrops(),
            BlockBehaviour.Properties.of().sound(SoundType.STONE).strength(5, 10).requiresCorrectToolForDrops()
    ));
    public static Material NICKEL = register(new Material(
            "nickel",
            RenderUtils.color(0xd4, 0xd4, 0xa9),
            "Ni",
            BlockTags.NEEDS_IRON_TOOL,
            BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.BASEDRUM).strength(5, 10).requiresCorrectToolForDrops(),
            BlockBehaviour.Properties.of().sound(SoundType.STONE).strength(5, 10).requiresCorrectToolForDrops()
    ));
    public static Material TITANIUM = register(new Material(
            "titanium",
            RenderUtils.color(0xcf, 0x71, 0xaf),
            "Ti",
            BlockTags.NEEDS_DIAMOND_TOOL,
            BlockBehaviour.Properties.of().instrument(NoteBlockInstrument.BASEDRUM).strength(5, 10).requiresCorrectToolForDrops(),
            BlockBehaviour.Properties.of().sound(SoundType.STONE).strength(5, 10).requiresCorrectToolForDrops()
    ));

    public static Material register(Material material) {
        MATERIALS.add(material);
        return material;
    }
}

File: src/main/kotlin/redcrafter07/processed/materials/Materials.kt
package redcrafter07.processed.materials

import redcrafter07.processed.gui.RenderUtils.color
import redcrafter07.processed.materials.data.MaterialBase
import java.util.stream.Stream

object Materials {
    val MATERIALS = ArrayList<MaterialBase>()

    val ALUMINIUM = register(MineOreMaterial("aluminium", color(0xd0, 0xd5, 0xd9)))
    val TIN = register(MineOreMaterial("tin", color(0xd5, 0xd6, 0xbc)))
    val NICKEL = register(MineOreMaterial("nickel", color(0x4d, 0xd4, 0xa9)))
    val TITANIUM = register(MineOreMaterial("titanium", color(0xcf, 0x71, 0xaf)))
    val URANIUM = register(MineOreMaterial("uranium", color(0x3c, 0xff, 0x49)))
    val NAQUADAH = register(MinerOreMaterial("naquadah", color(0x1f, 0x30, 0x21)))
    val IRON = register(IronMaterial)
    val COPPER = register(CopperMaterial)

    val STEEL = register(AlloyMaterial("steel", color(0x49, 0x4b, 0x4d)))
    val NICKEL_TITANIUM = register(AlloyMaterial("nickel_titanium", color(0x8f, 0xa3, 0xa9)))

    fun <T : MaterialBase> register(material: T) = material.apply { MATERIALS.add(this) }

    inline fun <reified T> getMaterials(): Stream<T> =
        MATERIALS.stream().mapMulti { v, consumer -> if (v is T) consumer.accept(v) }
}